### PR TITLE
WIP - CORS and fixes

### PR DESCRIPTION
**Pull request by raeesbhatti**
Conforms to: https://gist.github.com/ganeshvlrk/81e9f3fb838e90f7ef1c2bfdb08ca46a

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -565,18 +572,76 @@ class EGPlugin {
...
     this.printGatewayInfo()
-    this.printFunctions().then(() => this.printSubscriptions())
+    this.printEventTypes()
+      .then(() => this.printFunctions())
+      .then(() => this.printSubscriptions())
+      .then(() => this.printCORS())
+  }
+
+  async printShortDashboard () {
+    const lt = await this.client.listEventTypes()
```
Comment (mthenw) [inline src/index.js]: Please rename those two letters valiables with eventTypes, functions, subscriptions, cors. It's hard to follow them in the code.

File: src/index.js [hunk 2]
```diff
@@ -565,18 +572,76 @@ class EGPlugin {
...
-    this.serverless.cli.consoleLog(` space: ${this.client.config.space}`)
-    this.serverless.cli.consoleLog(` endpoint: ${this.client.config.eventsUrl}`)
+    this.serverless.cli.consoleLog(` Tenant: ${this.serverless.service.tenant}`)
+    this.serverless.cli.consoleLog(` App: ${this.serverless.service.app}`)
+    this.serverless.cli.consoleLog(` Domain: ${this.client.config.eventsUrl}`)
     this.serverless.cli.consoleLog('')
   }
 
+  printEventTypes () {
+    return this.client.listEventTypes().then(es => {
```
Comment (mthenw) [inline src/index.js]: `es` -> `eventTypes`

File: src/index.js [hunk 3]
```diff
@@ -565,18 +572,76 @@ class EGPlugin {
...
+      const table = new Table({
+        head: ['Name', 'Space', 'Authorizer'],
+        style: { head: ['bold'] }
+      })
+      es.forEach(x => {
```
Comment (mthenw) [inline src/index.js]: `x` -> `eventType`

File: src/index.js [hunk 4]
```diff
@@ -565,18 +572,76 @@ class EGPlugin {
...
+    fs.forEach(x => data.push({functionId: x.functionId}))
+    ss.forEach(x => {
+      const index = data.findIndex(i => i.functionId === x.functionId)
+      if (index > -1) {
+        const d = data[index]
+        return Object.assign(data[index], {eventType: [].concat(data[index].eventType || []).concat(x.eventType), method: [].concat(d.method || []).concat(x.method), path: x.path})
+      }
+      data.push({eventType: [x.eventType], functionId: x.functionId, method: [x.method], path: x.path})
+    })
+    cs.forEach(x => {
```
Comment (mthenw) [inline src/index.js]: This code is really hard to read. Completely not sure what's happening here. Can you extract it to smaller methods?

File: src/index.js [hunk 5]
```diff
@@ -599,14 +664,28 @@ class EGPlugin {
         style: { head: ['bold'] }
       })
       subscriptions.forEach(s => {
-        table.push([s.event || '', s.functionId || '', s.method || '', s.path || ''])
+        table.push([s.eventType || '', s.functionId || '', s.method || '', s.path || ''])
```
Comment (mthenw) [inline src/index.js]: I think it should also include subscription type.

File: src/index.js [hunk 6]
```diff
@@ -599,14 +664,28 @@ class EGPlugin {
...
   }
 
+  printCORS () {
+    return this.client.listCORS()
+      .then(cors => {
+        const table = new Table({
+          head: ['Method', 'Path', 'Origins', 'Methods', 'Headers', 'Allow Credentials'],
+          style: { head: ['bold'] }
+        })
+        cors.forEach(x => table.push([x.method || '', x.path || '', x.allowedOrigins.join(', ') || '', x.allowedMethods.join(', ') || '', x.allowedHeaders.join(', ') || '', JSON.stringify(x.allowCredentials)]))
```
Comment (mthenw) [inline src/index.js]: `cors` -> `corsConfigs`
`x` -> `cors` or `config`

File: src/index.js [hunk 7]
```diff
@@ -46,7 +46,13 @@ class EGPlugin {
           },
           dashboard: {
             usage: 'Show functions and subscriptions in Space for Event Gateway',
-            lifecycleEvents: ['dashboard']
+            lifecycleEvents: ['dashboard'],
+            options: {
+              event: {
```
Comment (mthenw) [inline src/index.js]: event?

**Comment by mthenw** [comment]
@raeesbhatti can you paste some example output for short and normal version?

**Comment by raeesbhatti** [comment]
Example output:
```
$ npx sls gateway dashboard --short

Event Gateway dashboard
┌────────────┬─────────────────────────────────┬─────────────┬───────────────────────────────────┬─────────────────┬─────────────────┬──────────────────────────────┬───────────────────┐
│ Event Type │ Function ID                     │ Method      │ Path                              │ Allowed Origins │ Allowed Methods │ Allowed Headers              │ Allow Credentials │
├────────────┼─────────────────────────────────┼─────────────┼───────────────────────────────────┼─────────────────┼─────────────────┼──────────────────────────────┼───────────────────┤
│ http.requ… │ eg-users-service-dev-getUser    │ GET, DELETE │ /raeesbhatti-myapp-dev/users2/:id │ *               │ HEAD, GET, POST │ Origin, Accept, Content-Type │ false             │
├────────────┼─────────────────────────────────┼─────────────┼───────────────────────────────────┼─────────────────┼─────────────────┼──────────────────────────────┼───────────────────┤
│ http.requ… │ eg-users-service-dev-deleteUser │ DELETE      │ /raeesbhatti-myapp-dev/users2/:id │                 │                 │                              │                   │
├────────────┼─────────────────────────────────┼─────────────┼───────────────────────────────────┼─────────────────┼─────────────────┼──────────────────────────────┼───────────────────┤
│ http.requ… │ eg-users-service-dev-createUser │ POST        │ /raeesbhatti-myapp-dev/users2     │ *               │ HEAD, GET, POST │ Origin, Accept, Content-Type │ false             │
└────────────┴─────────────────────────────────┴─────────────┴───────────────────────────────────┴─────────────────┴─────────────────┴──────────────────────────────┴───────────────────┘


$ npx sls gateway dashboard

Event Gateway

 Tenant: raeesbhatti
 App: myapp-dev
 Domain: https://raeesbhatti-myapp-dev.eventgateway-dev.io

Event Types
┌──────────────┬───────────────────────┬────────────┐
│ Name         │ Space                 │ Authorizer │
├──────────────┼───────────────────────┼────────────┤
│ http.request │ raeesbhatti-myapp-dev │ null       │
└──────────────┴───────────────────────┴────────────┘

Functions
┌─────────────────────────────────┬───────────┬────────────────────────────────────────────────────────────────────────────────┐
│ Function ID                     │ Region    │ ARN                                                                            │
├─────────────────────────────────┼───────────┼────────────────────────────────────────────────────────────────────────────────┤
│ eg-users-service-dev-getUser    │ us-east-1 │ arn:aws:lambda:us-east-1:211551830235:function:eg-users-service-dev-getUser    │
├─────────────────────────────────┼───────────┼────────────────────────────────────────────────────────────────────────────────┤
│ eg-users-service-dev-deleteUser │ us-east-1 │ arn:aws:lambda:us-east-1:211551830235:function:eg-users-service-dev-deleteUser │
├─────────────────────────────────┼───────────┼────────────────────────────────────────────────────────────────────────────────┤
│ eg-users-service-dev-createUser │ us-east-1 │ arn:aws:lambda:us-east-1:211551830235:function:eg-users-service-dev-createUser │
└─────────────────────────────────┴───────────┴────────────────────────────────────────────────────────────────────────────────┘

Subscriptions
┌──────────────┬──────┬─────────────────────────────────┬────────┬───────────────────────────────────┐
│ Event        │ Type │ Function ID                     │ Method │ Path                              │
├──────────────┼──────┼─────────────────────────────────┼────────┼───────────────────────────────────┤
│ http.request │ sync │ eg-users-service-dev-createUser │ POST   │ /raeesbhatti-myapp-dev/users2     │
├──────────────┼──────┼─────────────────────────────────┼────────┼───────────────────────────────────┤
│ http.request │ sync │ eg-users-service-dev-getUser    │ GET    │ /raeesbhatti-myapp-dev/users2/:id │
├──────────────┼──────┼─────────────────────────────────┼────────┼───────────────────────────────────┤
│ http.request │ sync │ eg-users-service-dev-deleteUser │ DELETE │ /raeesbhatti-myapp-dev/users2/:id │
└──────────────┴──────┴─────────────────────────────────┴────────┴───────────────────────────────────┘

CORS
┌────────┬───────────────────────────────────┬─────────┬─────────────────┬──────────────────────────────┬───────────────────┐
│ Method │ Path                              │ Origins │ Methods         │ Headers                      │ Allow Credentials │
├────────┼───────────────────────────────────┼─────────┼─────────────────┼──────────────────────────────┼───────────────────┤
│ POST   │ /raeesbhatti-myapp-dev/users2     │ *       │ HEAD, GET, POST │ Origin, Accept, Content-Type │ false             │
├────────┼───────────────────────────────────┼─────────┼─────────────────┼──────────────────────────────┼───────────────────┤
│ DELETE │ /raeesbhatti-myapp-dev/users2/:id │ *       │ HEAD, GET, POST │ Origin, Accept, Content-Type │ false             │
└────────┴───────────────────────────────────┴─────────┴─────────────────┴──────────────────────────────┴───────────────────┘
```

**Comment by ganeshvlrk** [comment]
@raeesbhatti Besides review, whats needed to have this merged? I dont see any tests added. Is that something that hasn't been merged?

File: src/index.js [hunk 8]
```diff
@@ -46,7 +46,13 @@ class EGPlugin {
           },
           dashboard: {
             usage: 'Show functions and subscriptions in Space for Event Gateway',
-            lifecycleEvents: ['dashboard']
+            lifecycleEvents: ['dashboard'],
+            options: {
+              short: {
+                usage: 'Short version of your Event Gateway dashboard',
```
Comment (mthenw) [inline src/index.js]: Can you add better explanation her? What does it mean "short"?

File: src/index.js [hunk 9]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+        'Path',
+        'Allowed Origins',
+        'Allowed Methods',
+        'Allowed Headers',
+        'Allow Credentials'
+      ],
+      style: { head: ['bold'] }
+    })
+    const data = []
+    functions.forEach((x) => data.push({ functionId: x.functionId }))
```
Comment (mthenw) [inline src/index.js]: `x` -> `func`

File: src/index.js [hunk 10]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+    subscriptions.forEach((x) => {
```
Comment (mthenw) [inline src/index.js]: `x`-> `subscription`

File: src/index.js [hunk 11]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
 
-  printGatewayInfo() {
-    this.serverless.cli.consoleLog(chalk.bold('Event Gateway'))
-    this.serverless.cli.consoleLog('')
-    this.serverless.cli.consoleLog(` space: ${this.client.config.space}`)
-    this.serverless.cli.consoleLog(` endpoint: ${this.client.config.eventsUrl}`)
-    this.serverless.cli.consoleLog('')
+  printDashboardInfo({ heading, content }) {
+    // eslint-disable-next-line no-console
+    console.log(`\n ${chalk.bold(heading)}`)
```
Comment (mthenw) [inline src/index.js]: Can we use `this.serverless...consoleLog` here?

File: src/index.js [hunk 12]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+    // eslint-disable-next-line no-console
+    console.log(`${content}\n`)
```
Comment (mthenw) [inline src/index.js]: Same as above.

File: src/index.js [hunk 13]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+        'Allowed Headers',
+        'Allow Credentials'
+      ],
+      style: { head: ['bold'] }
+    })
+    const data = []
+    functions.forEach((x) => data.push({ functionId: x.functionId }))
+    subscriptions.forEach((x) => {
+      // Check if another element with functionId as this subscription's functionId exists
+      const index = data.findIndex((i) => i.functionId === x.functionId)
```
Comment (mthenw) [inline src/index.js]: Hmm, maybe I'm missing something but the plugin registers only functions that have subscriptions. So I think we don't even need to fetch functions because we have functionID in subscription object. Basically, we don't need to make this detection here, we just assume if a function is used in a subscription it's registered.

File: src/index.js [hunk 14]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+      data.push({
+        eventType: [x.eventType],
+        functionId: x.functionId,
+        method: [x.method],
+        path: x.path
+      })
+    })
+    cors.forEach((x) => {
+      // Check if another element with exact path as this cors rule exists
+      const index = data.findIndex((i) => i.path === x.path)
```
Comment (mthenw) [inline src/index.js]: CORS is configured bot path and method.

File: src/index.js [hunk 15]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+      if (index > -1) {
```
Comment (mthenw) [inline src/index.js]: Also, I don't think we need to do this here. I think it would be simpler if we would find CORS configuration is subscriptions' `forEach` loop. Simple `cors.find()`.

File: src/index.js [hunk 16]
```diff
@@ -625,22 +632,141 @@ class EGPlugin {
...
+    return {
+      heading: 'Event Gateway dashboard',
+      content: table.toString()
+    }
+  }
+
+  getGatewayInfo() {
+    return {
+      heading: 'Event Gateway',
+      content:
```
Comment (mthenw) [inline src/index.js]: It should also support self-hosted EG, where there is only Space property. Is it possible to support that here?

File: test/index.js [hunk 17]
```diff
@@ -916,6 +916,78 @@ describe('Event Gateway Plugin', () => {
...
+                path: '/test',
+                method: 'GET'
+              }
+            }
+          ]
+        }
+      }
+    })
+
+    it('should return correct header information to print', async () => {
```
Comment (mthenw) [inline test/index.js]: Please add also tests for short dashboard and tests for a self-hosted EG when there is no tenant and app.

File: src/index.js (hunk 11 above)
Comment (raeesbhatti) [reply src/index.js]: No, since `this` was never bound to `printDashboardInfo`.

File: src/index.js (hunk 14 above)
Comment (raeesbhatti) [reply src/index.js]: Yes, but we're sorting the items by path only, and appending all methods to a single column.

File: src/index.js (hunk 15 above)
Comment (raeesbhatti) [reply src/index.js]: I'm not sure if I understand. The CORS logic here does not check for subscriptions, and they're returned independently from EG. This logic is simply about checking if another CORS rule exists in table, and adding to it if possible. The CORS rules are displayed separately from subscriptions

File: src/index.js (hunk 13 above)
Comment (raeesbhatti) [reply src/index.js]: But we need to get region and ARN for each of the functions.

File: src/index.js [hunk 18]
```diff
@@ -625,22 +633,142 @@ class EGPlugin {
...
+        (!this.serverless.service.tenant ? '' : `\nTenant: ${this.serverless.service.tenant}\n`) +
+        (!this.serverless.service.app ? '' : `App: ${this.serverless.service.app}\n`) +
+        (!this.serverless.service.space ? '' : `Space: ${this.serverless.service.space}\n`) +
```
Comment (mthenw) [inline src/index.js]: Space is not set on service level and also it should not be shown when app and tenant are set. https://github.com/serverless/serverless-event-gateway-plugin/blob/master/src/index.js#L638

File: src/index.js [hunk 19]
```diff
@@ -625,22 +633,142 @@ class EGPlugin {
...
 
-  printFunctions() {
+  getEventTypesInfo() {
+    return this.client.listEventTypes().then((eventTypes) => {
+      const table = new Table({
+        head: ['Name', 'Space', 'Authorizer'],
+        style: { head: ['bold'] }
+      })
+      eventTypes.forEach((eventType) =>
+        table.push([eventType.name || '', eventType.space || '', eventType.authorizerId || 'null'])
```
Comment (mthenw) [inline src/index.js]: There is always `name` and `space` for every event type. No need to make it conditional.

File: src/index.js [hunk 20]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
     return this.client.listSubscriptions().then((subscriptions) => {
       const table = new Table({
-        head: ['Event', 'Function ID', 'Method', 'Path'],
+        head: ['Event', 'Type', 'Function ID', 'Method', 'Path'],
         style: { head: ['bold'] }
       })
       subscriptions.forEach((s) => {
-        table.push([s.event || '', s.functionId || '', s.method || '', s.path || ''])
+        table.push([
+          s.eventType || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 21]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
+          s.type || '',
+          s.functionId || '',
+          s.method || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 22]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
-        head: ['Event', 'Function ID', 'Method', 'Path'],
+        head: ['Event', 'Type', 'Function ID', 'Method', 'Path'],
         style: { head: ['bold'] }
       })
       subscriptions.forEach((s) => {
-        table.push([s.event || '', s.functionId || '', s.method || '', s.path || ''])
+        table.push([
+          s.eventType || '',
+          s.type || '',
+          s.functionId || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 23]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
       const table = new Table({
-        head: ['Event', 'Function ID', 'Method', 'Path'],
+        head: ['Event', 'Type', 'Function ID', 'Method', 'Path'],
         style: { head: ['bold'] }
       })
       subscriptions.forEach((s) => {
-        table.push([s.event || '', s.functionId || '', s.method || '', s.path || ''])
+        table.push([
+          s.eventType || '',
+          s.type || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 24]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
+          s.functionId || '',
+          s.method || '',
+          s.path || ''
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 25]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
+      const table = new Table({
+        head: ['Method', 'Path', 'Origins', 'Methods', 'Headers', 'Allow Credentials'],
+        style: { head: ['bold'] }
       })
-      this.serverless.cli.consoleLog(chalk.bold('Subscriptions'))
-      this.serverless.cli.consoleLog(table.toString())
-      this.serverless.cli.consoleLog('')
+      corsConfigs.forEach((cors) =>
+        table.push([
+          cors.method || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 26]
```diff
@@ -649,24 +777,58 @@ class EGPlugin {
...
+          cors.path || '',
```
Comment (mthenw) [inline src/index.js]: it's required field

File: src/index.js [hunk 27]
```diff
@@ -625,22 +633,142 @@ class EGPlugin {
...
+        'Allow Credentials'
+      ],
+      style: { head: ['bold'] }
+    })
+    const data = []
+    functions.forEach((func) => data.push({ functionId: func.functionId }))
+    subscriptions.forEach((sub) => {
+      // Check if another element with functionId as this subscription's functionId exists
+      const index = data.findIndex((i) => i.functionId === sub.functionId)
+      if (index > -1) {
```
Comment (mthenw) [inline src/index.js]: Even if we need to get region and ARN it would be more effective to loop on subscriptions and for every subscription just find the corresponding function instead of modifying existing entries. Same applies for CORS and event types.

For every subscription it should:
1. find function
2. find event type
3. find CORS
4 put everything into `data`

Comment (raeesbhatti) [reply src/index.js]: What if a function doesn't have subscriptions?

Comment (mthenw) [reply src/index.js]: Oh yeah, a function can be an authorizer. Looking at the expected output I think that we should loop over functions and then find subscriptions, then event types (based on subscriptions) and CORS (based on subscriptions).

Comment (mthenw) [reply src/index.js]: It's complicated too much and it's hard to read this code even though it's pretty simple functionality. Currently, it works like that:
1. iterate over functions
    1. add information about a function to the output
1. iterate over subscriptions
    1. find the function in the output
    1. modify info about even types, methods and paths
1. iterate over cors configs 
    1. find the path in the output 
    1. modify info about cors
1. iterate over event types
    1. find the event type in the output
    1. modify info about authorizer

It would be simpler if that would be like this:
1. iterate over function, for each function:
    1. find subscriptions with the same function ID
    1. add info about event types, methods and path
    1. find infor about cors
    1. add info about cors
    1. find info about event types
    1. add info about authorizer

The latter approach would eliminate the need for those `if (index > -1) {` (modifying existing records) and it can be nicely implemented with map/reduce functions.

File: src/index.js [hunk 28]
```diff
@@ -625,22 +633,144 @@ class EGPlugin {
...
+    }
+  }
+
+  getGatewayInfo() {
+    return {
+      heading: 'Event Gateway',
+      content:
+        (!this.serverless.service.tenant ? '' : `\nTenant: ${this.serverless.service.tenant}\n`) +
+        (!this.serverless.service.app ? '' : `App: ${this.serverless.service.app}\n`) +
+        (!this.serverless.service.custom.eventgateway.space
```
Comment (alexdebrie) [inline src/index.js]: This throws an error if I don't have an `eventgateway` section in the `custom` block of my `serverless.yml`. Can you fix?
